azure-quantum/azure/quantum/job/base_job.py

```diff
@@ -0,0 +1,240 @@
+##
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+##
+import abc
+import logging
+import uuid
+
+from urllib.parse import urlparse
+from typing import Any, Dict, TYPE_CHECKING
+from urllib.parse import urlparse
+from azure.storage.blob import BlobClient
+
+from azure.quantum.storage import create_container_using_client, get_container_uri, upload_blob, download_blob, ContainerClient
+from azure.quantum._client.models import JobDetails
+
+
+if TYPE_CHECKING:
+    from azure.quantum.workspace import Workspace
+
+
+logger = logging.getLogger(__name__)
+
+DEFAULT_TIMEOUT = 100
+
+
+class BaseJob(abc.ABC):
+    # Optionally override these to create a Provider-specific Job subclass
+    input_data_format = None
+    output_data_format = None
+    provider_id = None
+
+    """
+    Base job class with methods to create a job from raw blob data,
+    upload blob data and download results.
+    """
+    @staticmethod
+    def create_job_id() -> str:
+        """Create a unique id for a new job."""
+        return str(uuid.uuid1())
+
+    @property
+    def container_name(self):
+        return f"job-{self.id}"
+
+    @classmethod
+    def from_input_data(
+        cls,
+        workspace: "Workspace",
+        name: str,
```

Comment:
should this be just `Workspace` (no quotes?)

Reply:
I made this a string to avoid a circular import.